### PR TITLE
Round service

### LINKED ISSUE #56: Implement Round Service
- [x] enterPlayersChooseState
- [x] playersChoosePunchlines
- [X] enterHostChoosesState
- [X] hostChoosesWinner

### PR DESCRIPTION
**Pull request by taitfuller**
Implements Round service containing the majority of our game logic. Includes the following services:

-  enterPlayersChooseState
-  playersChoosePunchlines
-  enterHostChoosesState
-  hostChoosesWinner

Closes #56

### CODE REVIEW

File: backend/src/services/round.service.ts [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+  const round = game.rounds.slice(-1)[0];
+
+  if (
+    round !== undefined &&
+    round.state === RoundState.before &&
+    round.host.toString() === playerId
+  ) {
+    round.state = RoundState.playersChoose;
+
+    // TODO: Allocate player cards here?
```
Comment (Alxjlv) [inline backend/src/services/round.service.ts:20]: Do you mean players drawing a new card here?

File: backend/src/services/round.service.ts [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+    !round.punchlinesByPlayer.has(playerId) &&
+    // Check the number of punchlines chosen matches the setup type
+    ((round.setup.type === SetupType.pickOne &&
+      chosenPunchlines.length === 1) ||
+      (round.setup.type === SetupType.pickTwo &&
+        chosenPunchlines.length === 2) ||
+      (round.setup.type === SetupType.drawTwoPickThree &&
+        chosenPunchlines.length === 3)) &&
+    player !== null &&
+    // Player has chosen punchlines in their hand
```
Comment (Alxjlv) [inline backend/src/services/round.service.ts:54]: Liking the comments!

File: backend/src/services/round.service.ts [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+): Promise<{ playerId: Player["id"]; punchlines: string[] }> => {
+  const game = await getGame(gameCode);
+  const round = game.rounds.slice(-1)[0];
+
+  if (
+    round !== undefined &&
+    round.state === RoundState.hostChooses &&
+    round.host.toString() === playerId
+  ) {
+    const winningEntry = Array.from(round.punchlinesByPlayer.entries()).find(
```
Comment (Alxjlv) [inline backend/src/services/round.service.ts:103]: I see why you renamed the method now

**Review by Alxjlv — APPROVED** [review]
Solid work! Sorry for the delay

File: backend/src/services/round.service.ts (hunk 1 above)
Comment (taitfuller) [reply backend/src/services/round.service.ts:20]: Yeah although I think it is best to have the allocation function in the game service, it's just a better of where and when that function is called from & it has multiple answers

**Pull request merged**